_episodes/04-networks-are-like-onions.md

```diff
@@ -0,0 +1,529 @@
+    Total params: 307,300
+    Trainable params: 307,300
+    Non-trainable params: 0
+    _________________________________________________________________
+> > ~~~
+> > {: .output}
+> {: .solution}
+{: .challenge}
+
+
+
+Of course, we can decrease the number of units in our hidden layer, but this also decreases the number of patterns our network can remember. Moreover, if we increase the image size, this number of weights will explode, even though the task of recognizing large images is not necessarily more difficult than the task of recognizing small images.
+
+The solution is that we make the network learn in a 'smart' way. The features that we learn should be similar both for small and large images, and similar features (e.g. edges, corners) can appear anywhere in the image (in mathematical terms: *translation invariant*). We do this by making use of a concepts from image processing that precede Deep Learning.
+
+A **convolution matrix**, or **kernel**, is a matrix transformation that we 'slide' over the image to calculate features at each position of the image. For each pixel, we calculate the matrix product between the kernel and the pixel with its surroundings. A kernel is typically small, between 3x3 and 7x7 pixels. We can for example think of the simple 3x3 kernel:
+~~~
+[[0, 0, 0,],
+ [1, 1, 1],
+ [0, 0, 0]]
+~~~
+{: .language-python}
+This kernel will give a high value to a pixel if it's in the middle of a horizontal edge.
+Note that for RGB images, the kernel should also have a depth of 3.
+
+In our **convolutional layer** our hidden units are a number of convolutional matrices (or kernels), where the values of the matrices are the weights that we learn in the training process. The output of a convolutional layer is an 'image' for each of the kernels, that gives the output of the kernel applied to each pixel.
+
+> ## Border pixels
+>
+> What, do you think, happens to the border pixels when applying a convolution?
+>
+> > ## Solution
+> >
+> > There are different ways of dealing with border pixels.
+> > You can ignore them, which means that your output image is slightly smaller then your input.
+> > It is also possible to 'pad' the borders, e.g. with the same value or with zeros, so that the convolution can also be applied to the border pixels.
+> > In that case, the output image will have the same size as the input image.
+> {: .solution}
+{: .challenge}
+
+> ## Number of model parameters
+>
+> Suppose we apply a convolutional layer with 100 kernels of size 3 * 3 * 3 (the last dimension applies to the rgb channels) to our images of 32 * 32 * 3 pixels. How many parameters do we have? Assume, for simplicity, that the kernels do not use bias terms. Compare this to the answer of the previous exercise
+>
+> > ## Solution
+> >
+> > We have 100 matrices with 3 * 3 * 3 = 27 values each so that gives 27 * 100 = 2700 weights. This is a magnitude of 100 less than the fully connected layer with 100 units!
+> {: .solution}
+{: .challenge}
+
+
```

Comment:
I would cut the lesson here. We should try to focus on the exercises at this point and perhaps "train" an entire network with convolutions for the fun of it. I also end my review here.

Reply:
I kinda agree that training a NN with convolutions here would be nice.
Introduce a concept and use it, then move on to the next concept.

I do feel that the drop out is very useful to have in this lesson as it helps a lot against overfitting.

Reply:
But would you suggest then to not introduce the pooling layers? Or only introduce them after we first train a CNN without pooling layers? 
I think CNNs almost always include some sort of pooling and they are easier to understand than convolutional layers, so I prefer to keep them in the same episode. 

But I will try to add an exercise in between, before we introduce the pooling layer.